File: CoseSign1.Transparent.MST/CborProblemDetails.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

namespace CoseSign1.Transparent.MST;

using System;
using System.Collections.Generic;
using System.Formats.Cbor;

/// <summary>
/// Represents parsed CBOR problem details from RFC 9290 (Concise Problem Details).
/// </summary>
/// <remarks>
/// RFC 9290 defines standard fields for problem details in CBOR format,
/// returned by Azure Code Transparency Service with Content-Type:
/// <c>application/concise-problem-details+cbor</c>.
///
/// Standard CBOR integer keys:
/// <list type="table">
///   <item><term>-1</term><description>type (URI reference)</description></item>
///   <item><term>-2</term><description>title (human-readable summary)</description></item>
///   <item><term>-3</term><description>status (HTTP status code)</description></item>
///   <item><term>-4</term><description>detail (human-readable explanation)</description></item>
///   <item><term>-5</term><description>instance (URI reference for the occurrence)</description></item>
/// </list>
///
/// String keys (<c>"type"</c>, <c>"title"</c>, etc.) are also accepted for interoperability.
/// </remarks>
public class CborProblemDetails
{
    /// <summary>
    /// Gets or sets the problem type URI reference (CBOR key: -1 or "type").
    /// </summary>
    public string? Type { get; set; }

    /// <summary>
    /// Gets or sets a short human-readable summary of the problem (CBOR key: -2 or "title").
    /// </summary>
    public string? Title { get; set; }

    /// <summary>
    /// Gets or sets the HTTP status code (CBOR key: -3 or "status").
    /// </summary>
    public int? Status { get; set; }

    /// <summary>
    /// Gets or sets a human-readable explanation of the problem (CBOR key: -4 or "detail").
    /// </summary>
    public string? Detail { get; set; }

    /// <summary>
    /// Gets or sets a URI reference for the specific occurrence (CBOR key: -5 or "instance").
    /// </summary>
    public string? Instance { get; set; }

    /// <summary>
    /// Gets or sets additional extension fields not covered by the standard keys.
    /// </summary>
    public Dictionary<string, object?>? Extensions { get; set; }

    /// <summary>
    /// Parses CBOR-encoded problem details (RFC 9290) from a byte array.
    /// </summary>
    /// <param name="cborBytes">The CBOR-encoded problem details.</param>
    /// <returns>The parsed problem details, or <c>null</c> if parsing fails.</returns>
    public static CborProblemDetails? TryParse(byte[] cborBytes)
    {
        if (cborBytes == null || cborBytes.Length == 0)
        {
            return null;
        }

        try
        {
            var reader = new CborReader(cborBytes);
            return ParseFromReader(reader);
        }
        catch (CborContentException)
        {
            return null;
        }
        catch (InvalidOperationException)
        {
            return null;
        }
    }

    /// <summary>
    /// Parses CBOR problem details from a <see cref="CborReader"/> positioned at a map.
    /// </summary>
    private static CborProblemDetails? ParseFromReader(CborReader reader)
    {
        if (reader.PeekState() != CborReaderState.StartMap)
        {
            return null;
        }

        var details = new CborProblemDetails();
        var extensions = new Dictionary<string, object?>();

        reader.ReadStartMap();

        while (reader.PeekState() != CborReaderState.EndMap)
        {
            var keyState = reader.PeekState();

            if (keyState == CborReaderState.NegativeInteger || keyState == CborReaderState.UnsignedInteger)
            {
                int key = reader.ReadInt32();
                switch (key)
                {
                    case -1: details.Type = ReadStringValue(reader); break;
                    case -2: details.Title = ReadStringValue(reader); break;
                    case -3: details.Status = ReadIntValue(reader); break;
                    case -4: details.Detail = ReadStringValue(reader); break;
                    case -5: details.Instance = ReadStringValue(reader); break;
                    default: extensions[$"key_{key}"] = ReadAnyValue(reader); break;
                }
            }
            else if (keyState == CborReaderState.TextString)
            {
                string key = reader.ReadTextString();
                switch (key.ToLowerInvariant())
                {
                    case "type": details.Type = ReadStringValue(reader); break;
                    case "title": details.Title = ReadStringValue(reader); break;
                    case "status": details.Status = ReadIntValue(reader); break;
                    case "detail": details.Detail = ReadStringValue(reader); break;
                    case "instance": details.Instance = ReadStringValue(reader); break;
                    default: extensions[key] = ReadAnyValue(reader); break;
                }
            }
            else
            {
                reader.SkipValue();
                reader.SkipValue();
            }
        }

        reader.ReadEndMap();

        if (extensions.Count > 0)
        {
            details.Extensions = extensions;
        }

        return details;
    }

    private static string? ReadStringValue(CborReader reader)
    {
        if (reader.PeekState() == CborReaderState.TextString)
        {
            return reader.ReadTextString();
        }

        if (reader.PeekState() == CborReaderState.Null) { reader.ReadNull(); return null; }
        reader.SkipValue();
        return null;
    }

    private static int? ReadIntValue(CborReader reader)
    {
        var state = reader.PeekState();
        if (state == CborReaderState.UnsignedInteger || state == CborReaderState.NegativeInteger)
        {
            return reader.ReadInt32();
        }

        if (state == CborReaderState.Null) { reader.ReadNull(); return null; }
        reader.SkipValue();
        return null;
    }

    private static object? ReadAnyValue(CborReader reader)
    {
        return reader.PeekState() switch
        {
            CborReaderState.TextString => reader.ReadTextString(),
            CborReaderState.ByteString => reader.ReadByteString(),
            CborReaderState.UnsignedInteger => reader.ReadInt64(),
            CborReaderState.NegativeInteger => reader.ReadInt64(),
            CborReaderState.Boolean => reader.ReadBoolean(),
            CborReaderState.Null => ReadNull(reader),
            CborReaderState.SinglePrecisionFloat => reader.ReadSingle(),
            CborReaderState.DoublePrecisionFloat => reader.ReadDouble(),
            CborReaderState.HalfPrecisionFloat => reader.ReadDouble(),
            _ => Skip(reader),
        };
    }

    private static object? ReadNull(CborReader reader) { reader.ReadNull(); return null; }
    private static object? Skip(CborReader reader) { reader.SkipValue(); return null; }

    /// <summary>
    /// Returns a human-readable summary of the problem details.
    /// </summary>
    public override string ToString()
    {
        var parts = new List<string>();
        if (!string.IsNullOrEmpty(Title))
        {
            parts.Add($"Title: {Title}");
        }

        if (Status.HasValue)
        {
            parts.Add($"Status: {Status}");
        }

        if (!string.IsNullOrEmpty(Detail))
        {
            parts.Add($"Detail: {Detail}");
        }

        if (!string.IsNullOrEmpty(Type))
        {
            parts.Add($"Type: {Type}");
        }

        if (!string.IsNullOrEmpty(Instance))
        {
            parts.Add($"Instance: {Instance}");
        }

        return parts.Count > 0 ? string.Join(", ", parts) : "No details available";
    }
}

File: native/rust/extension_packs/mst/client/src/cbor_problem_details.rs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

//! RFC 9290 CBOR Problem Details parser.
//!
//! Parses structured error bodies returned by the Azure Code Transparency Service
//! with Content-Type `application/concise-problem-details+cbor`.

use cbor_primitives::CborDecoder;
use std::collections::HashMap;
use std::fmt;

/// Parsed CBOR problem details per RFC 9290.
///
/// Standard CBOR integer keys:
/// - `-1` → type (URI reference)
/// - `-2` → title (human-readable summary)
/// - `-3` → status (HTTP status code)
/// - `-4` → detail (human-readable explanation)
/// - `-5` → instance (URI reference for the occurrence)
///
/// String keys (`"type"`, `"title"`, etc.) are also accepted for interoperability.
#[derive(Debug, Clone, Default)]
pub struct CborProblemDetails {
    /// Problem type URI reference (CBOR key: -1 or "type").
    pub problem_type: Option<String>,
    /// Short human-readable summary (CBOR key: -2 or "title").
    pub title: Option<String>,
    /// HTTP status code (CBOR key: -3 or "status").
    pub status: Option<i64>,
    /// Human-readable explanation (CBOR key: -4 or "detail").
    pub detail: Option<String>,
    /// URI reference for the specific occurrence (CBOR key: -5 or "instance").
    pub instance: Option<String>,
    /// Additional extension fields not covered by the standard keys.
    pub extensions: HashMap<String, String>,
}

impl CborProblemDetails {
    /// Attempts to parse CBOR problem details from a byte slice.
    ///
    /// Returns `None` if the bytes are empty or cannot be parsed as a CBOR map.
    pub fn try_parse(cbor_bytes: &[u8]) -> Option<Self> {
        if cbor_bytes.is_empty() {
            return None;
        }
        Self::parse_inner(cbor_bytes)
    }

    fn parse_inner(cbor_bytes: &[u8]) -> Option<Self> {
        let mut d = cose_sign1_primitives::provider::decoder(cbor_bytes);
        let map_len = d.decode_map_len().ok()?;
        let count = map_len.unwrap_or(0);

        let mut details = CborProblemDetails::default();

        for _ in 0..count {
            // Peek at the key type to decide how to decode it
            let key_type = d.peek_type().ok();
            match key_type {
                Some(cbor_primitives::CborType::NegativeInt)
                | Some(cbor_primitives::CborType::UnsignedInt) => {
                    let neg_key = d.decode_i64().ok()?;
                    match neg_key {
                        -1 => details.problem_type = d.decode_tstr().ok().map(|s| s.to_string()),
                        -2 => details.title = d.decode_tstr().ok().map(|s| s.to_string()),
                        -3 => details.status = d.decode_i64().ok(),
                        -4 => details.detail = d.decode_tstr().ok().map(|s| s.to_string()),
                        -5 => details.instance = d.decode_tstr().ok().map(|s| s.to_string()),
                        _ => {
                            let val = d
                                .decode_tstr()
                                .ok()
                                .map(|s| s.to_string())
                                .unwrap_or_default();
                            details.extensions.insert(format!("key_{}", neg_key), val);
                        }
                    }
                }
                Some(cbor_primitives::CborType::TextString) => {
                    let str_key = match d.decode_tstr().ok() {
                        Some(s) => s.to_string(),
                        None => break,
                    };
                    let str_key_lower = str_key.to_lowercase();
                    match str_key_lower.as_str() {
                        "type" => {
                            details.problem_type = d.decode_tstr().ok().map(|s| s.to_string())
                        }
                        "title" => details.title = d.decode_tstr().ok().map(|s| s.to_string()),
                        "status" => details.status = d.decode_i64().ok(),
                        "detail" => details.detail = d.decode_tstr().ok().map(|s| s.to_string()),
                        "instance" => {
                            details.instance = d.decode_tstr().ok().map(|s| s.to_string())
                        }
                        _ => {
                            let val = d.decode_tstr().ok().map(|s| s.to_string());
                            if let Some(v) = val {
                                details.extensions.insert(str_key, v);
                            }
                        }
                    }
                }
                _ => break,
            }
        }

        Some(details)
    }
}

impl fmt::Display for CborProblemDetails {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut parts = Vec::new();
        if let Some(ref title) = self.title {
            parts.push(format!("Title: {}", title));
        }
        if let Some(status) = self.status {
            parts.push(format!("Status: {}", status));
        }
        if let Some(ref detail) = self.detail {
            parts.push(format!("Detail: {}", detail));
        }
        if let Some(ref t) = self.problem_type {
            parts.push(format!("Type: {}", t));
        }
        if let Some(ref inst) = self.instance {
            parts.push(format!("Instance: {}", inst));
        }
        if parts.is_empty() {
            write!(f, "No details available")
        } else {
            write!(f, "{}", parts.join(", "))
        }
    }
}
